### PR TITLE
Add pip installer

### LINKED ISSUE #1642: Workflow linter to analyse PyPI library
Would share significant codebase with:
- https://github.com/databrickslabs/ucx/issues/1640

### PR DESCRIPTION
**Pull request by JCZuurmond**
## Changes
Add a pip installer. 

@ericvergnaud and I concluded to separate the installing of libraries from register/linting them. An installer makes a library available to the path look up by i) installing the library and ii) augmenting the path lookup.

### Linked issues
Resolves #1642

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 1]
```diff
@@ -63,50 +63,53 @@ def __init__(self, ws: WorkspaceClient, task: jobs.Task):
...
             yield DependencyProblem('not-yet-implemented', 'Requirements library is not yet implemented')
 
+    def _register_task_dependencies(self, graph: DependencyGraph) -> Iterable[DependencyProblem]:
+        yield from self._register_notebook(graph)
+        yield from self._register_spark_python_task(graph)
+        yield from self._register_python_wheel_task(graph)
+        yield from self._register_spark_jar_task(graph)
+        yield from self._register_run_job_task(graph)
+        yield from self._register_pipeline_task(graph)
+        yield from self._register_existing_cluster_id(graph)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/jobs.py:110]: Move to install libraries

File: src/databricks/labs/ucx/source_code/graph.py [hunk 2]
```diff
@@ -44,6 +46,10 @@ def dependency(self):
     def path(self):
         return self._dependency.path
 
+    def install_library(self, library: str) -> list[DependencyProblem]:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/graph.py:49]: @ericvergnaud : Do we want to add this to the graph? We could also add it to the path lookup as we are not doing anything with the graph.

File: tests/unit/source_code/test_site_packages.py [hunk 3]
```diff
@@ -1,7 +1,30 @@
-from databricks.labs.ucx.source_code.site_packages import SitePackages
+from pathlib import Path
+
+
+from databricks.labs.ucx.source_code.site_packages import PipInstaller, SitePackages
 from tests.unit import locate_site_packages
 
 
+def test_pip_installer_resolve_library(mock_path_lookup):
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_site_packages.py:7]: should be named ...resolves_library ?

**Comment by JCZuurmond** [comment]
Rebasing went wrong, creating new PR

**Pull request closed without merging**